text/0070-pd-record-historical-hot-region.md

```diff
@@ -0,0 +1,163 @@
+  | TABLE_NAME     | varchar(64) | YES  |      | NULL    |       |
+  | INDEX_NAME     | varchar(64) | YES  |      | NULL    |       |
+  | REGION_ID      | bigint(21)  | YES  |      | NULL    |       |
+  | TYPE           | varchar(64) | YES  |      | NULL    |       |
+  | MAX_HOT_DEGREE | bigint(21)  | YES  |      | NULL    |       |
+  | FLOW_BYTES     | bigint(21)  | YES  |      | NULL    |       |
+  // New fields
+  | KEY_RATE       | bigint(21)  | YES  |      | NULL    |       |
+  | QUERY_RATE     | bigint(21)  | YES  |      | NULL    |       |
+  | STORE_ID       | bigint(21)  | YES  |      | NULL    |       |
+  | UPDATE_TIME    | datetime    | YES  |      | NULL    |       |
+  // Deleted fields
+  | REGION_COUNT   | bigint(21)  | YES  |      | NULL    |       |
+  +----------------+-------------+------+------+---------+-------+
+  ```
+  * Add `STORE_ID` to track the machine of region.
+  * Add `UPDATE_TIME` to support history.
+  * Add `KEY_RATE` and `QUERY_RATE` for future expansion in hotspot determination dimensions.
+  * Remove `REGION_COUNT` for disuse and repeat with `STORE_ID`.
+
+2. Data size estimation
+
+    one record size: 8 * 8B(bitint) + 1 * 8B(datetime) + 4 * 64B(varchar(64)) = 328B
+    Below table show data size per day and per month in 5,10,15 minutes record interval respectively given the maximum number of hotspot regions  is 1000:
+
+  | Record Length (B)| Time Interval (Min)  | Data Size Per Day (MB)   | Data Size Per Month (MB)   |
+  | ---------------- | -------------------- | ------------------------ | -------------------------- |
+  | 328              | 5                    | 90.08789063              | 2702.636719                |
+  | 328              | 10                   | 45.04394531              | 1351.318359                |
+  | 328              | 15                   | 30.02929688              | 900.8789063                |
+
+### Design in PD
+1. Timing write：
+
+     The leader of PD will periodically encrypt  `start_key` and `end_key`  in data,and write hotspot region data into `LevelDB`.The write interval can be configured. The write fieldes are: `region_id`, `type`,  `max_hot_degree`, `flow_bytes`, `key_rate`, `query_rate`, `store_id`, `update_time`, `start_key`,  `end_key`.
+
+2. Timing delete
+
+   PD runs the delete check task periodically, and deletes the hotspot region data that exceeds the configured TTL time.
+
+3. Pull interface
+
+   PD query the data in `LevleDB` with filters pushed down, decodes the data and returns to TiDB.
+
+4. New options
+
+   There are two config options need to be add in PD’s `config.go`:
+
+   * `HisHotRegionSaveInterval`:  time interval for pd to record hotspot region information, default: 15 minutes.
+   * `HisHotRegionTTL`: maximum hold day for his hot region, default: 30 days.
+
+### Design in TiDB
+
+1. Add memory table `TIDB_HOT_REGIONS_HISTORY`：
+
+   Create a new memory table `TIDB_HOT_REGIONS_HISTORY` to the existing `INFORMATION_SCHEMA` database.
```

Comment:
I think the gc policy should consider memory consuming. Still I think memory table may not be a good choice especially in big cluster cases, the hot regions count might be large and each tidb-server maintained `TIDB_HOT_REGIONS_HISTORY` while we just only need 1 replication.

Reply:
it not table in memory,All data will be stored in pd's leveldb for persistent storage.In TiDB, TIDB_HOT_REGIONS_HISTORY is a virtual table which construct by data in pd